### PR TITLE
Implement propagation of IP routes into apps via DHCP

### PR DESCRIPTION
**Pull request by milan-zededa**
Most of the edge deployments will deploy applications (VMs or containers) that will have
connectivity to both WAN (Internet) and LAN (shop floor, machine floor, intranet).
There may be a single WAN interface and multiple LAN interfaces. With the IP route
configurability of Network instances implemented by this PR (and APIs extensions already
added [here](https://github.com/lf-edge/eve-api/pull/38)), we can automatically (zero-touch) provision routing for applications that have
access to both the Internet and one or more LANs.
This is now possible because EVE allows to:
 - use DHCP to propagate the connected routes to applications (routes for external networks 
   that uplink ports are connected to)
 - configure a set of static IP routes for a network instance and have them propagated
   to applications

Another common case is using one application as a network gateway for other applications
running on the same device. The gateway application may provide some network function(s),
such as firewall, IDS, network monitoring, etc. Such application will connect on one side with
the external network(s) using directly attached network adapter(s) or via switch network
instance(s), and the other side will make use of an air-gap local network instance to connect
with other applications running on the device. Propagated static IP routes are necessary
to make the application traffic flow through the gateway app. In theory, multiple network
functions can be chained together in this way using several air-gap network instances
with static IP routes.

PR with examples for eden: https://github.com/lf-edge/eden/pull/938
They allow to easily and quickly perform manual testing of this functionality.
Eventually, I would like to prepare automated tests, but for now examples will do.

### CODE REVIEW

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 1]
```diff
@@ -419,6 +419,34 @@ func parseDnsNameToIpList(
...
+	}
+	gatewayIP := net.ParseIP(route.Gateway)
+	if gatewayIP == nil {
+		return errors.New("gateway IP address is invalid")
+	}
+	if gatewayIP.IsUnspecified() {
+		return errors.New("gateway IP address is all-zeroes")
+	}
+	if gatewayIP.Equal(config.Gateway) {
+		return errors.New("static route gateway is the same as the network instance gateway")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:441]: Why is this an error? Does the gateway always have to be external to the device and network instance? 
If so, shouldn't the check be that the gateway doesn't fall in the subnet assigned to the network instance?

File: pkg/pillar/nireconciler/genericitems/dnsmasq.go [hunk 2]
```diff
@@ -606,24 +626,57 @@ func (c *DnsmasqConfigurator) CreateDnsmasqConfig(buffer io.Writer, dnsmasq Dnsm
...
-					return writeErr(err)
-				}
-			}
+			// From RFC 3442:
+			// If the DHCP server returns both a Classless Static Routes option and
+			// a Router option, the DHCP client MUST ignore the Router option.
+			//
+			// This means that we have to include a default "classless" route
+			// even though the Router option is provided.
+			// We keep the Router option for clients that do not support the Classless
```
Comment (eriknordmark) [inline pkg/pillar/nireconciler/genericitems/dnsmasq.go:673]: Makes sense to test this with a range of guest VMs including some version(s) of Windows, to make sure no client is upset by seeing both options.

File: pkg/pillar/nireconciler/linux_config.go [hunk 3]
```diff
@@ -636,6 +636,40 @@ func (r *LinuxNIReconciler) getIntendedNIL3Cfg(niID uuid.UUID) dg.Graph {
 			}, nil)
 		}
 	}
+	// Add statically defined routes into the NI routing table.
+	// No need to do this if GW is one of the apps and traffic from one app
```
Comment (eriknordmark) [inline pkg/pillar/nireconciler/linux_config.go:640]: If 10.0.0.0/16 is routed out one NI and 10.0.2.0/24 is routed to the default NI's bridgeIP or a host on the default subnet, then you can't supress the /24 route, can you?

File: pkg/pillar/nireconciler/linux_config.go [hunk 4]
```diff
@@ -636,6 +636,40 @@ func (r *LinuxNIReconciler) getIntendedNIL3Cfg(niID uuid.UUID) dg.Graph {
...
+	// to another is just forwarded by the host, not routed.
+	for _, route := range ni.config.StaticRoutes {
+		isAppGW := r.getNISubnet(ni).Contains(route.Gateway)
+		if isAppGW && r.disableAllOnesNetmask {
+			// Route is not needed inside the host, traffic is just forwarded
+			// by the bridge.
+			continue
+		}
+		if !isAppGW && !r.routeGwIsConnected(route, ni.bridge.Uplink) {
```
Comment (eriknordmark) [inline pkg/pillar/nireconciler/linux_config.go:649]: I guess there is an unstated assumption about the routes only being applicable to a particular NI, which I think we must document.

If I have an app instance connected to two NIs; the default local NI and a switch NI out eth1, then I might assume I can use EVE to publish a route on the first where the gateway is on the seconnd NI. AFAIK a guest VM like Linux would accept such a route (unless there is something special in some dhcp client code to tie to received routes to the interface on which the dhcp client received the option.)

File: pkg/pillar/nireconciler/linux_config.go [hunk 5]
```diff
@@ -781,22 +815,20 @@ func (r *LinuxNIReconciler) getIntendedDnsmasqCfg(niID uuid.UUID) (items []dg.It
...
+	// dnsmasq advertises a router (default route) for network instance, unless:
+	//  a) network instance is air-gapped (without uplink)
+	//  b) uplink is app-shared and without default route
+	var gatewayIP net.IP
+	if bridgeIP != nil {
+		gatewayIP = bridgeIP.IP
+	}
+	var withDefaultRoute bool
+	airGap := ni.bridge.Uplink.IfName == ""
+	if !airGap && (ni.bridge.Uplink.IsMgmt || r.niHasDefRoute(ni)) {
```
Comment (eriknordmark) [inline pkg/pillar/nireconciler/linux_config.go:827]: What's the need to check IsMgmt? Is this for a case when the management interface does not have a default route but somehow received all necessary prefix routes to reach the controller etc?

Note that the comment on L820 doesn't talk about IsMgmt.

File: pkg/pillar/nireconciler/linux_config.go [hunk 6]
```diff
@@ -810,17 +842,64 @@ func (r *LinuxNIReconciler) getIntendedDnsmasqCfg(niID uuid.UUID) (items []dg.It
 		ntpServers = append(ntpServers, ni.config.NtpServer)
 	}
 	ntpServers = generics.FilterDuplicatesFn(ntpServers, netutils.EqualIPs)
+	var propagateRoutes []types.IPRoute
```
Comment (eriknordmark) [inline pkg/pillar/nireconciler/linux_config.go:845]: Why do you do this work even if PropagateConnRoutes is not set?

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 1 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:441]: Gateway can be either external or one of the applications. So gateway inside the network instance subnet is OK.
However, I'm not sure if it would be wise to let the user create IP routes with the NI bridge as the gateway. Traffic routed by such route would arrive to EVE and then continue with the default route of the uplink, regardless of which uplink port is being used by NI at that moment. I'm not sure if we want to allow it. I though that it would be better to require that user enters specific external (or app) gateway.

File: pkg/pillar/nireconciler/genericitems/dnsmasq.go (hunk 2 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/genericitems/dnsmasq.go:673]: Will do

File: pkg/pillar/nireconciler/linux_config.go (hunk 3 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:640]: Here we only suppress installation of routes inside the NI routing table _on the host_ when traffic is only forwarded by the bridge, not routed (all-ones netmask is disabled and route is between apps on the same subnet). In that case, the route would not be applied anyway. But we still propagate all the routes to the applications using DHCP.

File: pkg/pillar/nireconciler/linux_config.go (hunk 4 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:649]: Yes, routes configured for a given NI should only apply to that network instance and interfaces connected to it. I will make sure to document this.

I believe that with this restriction it leads to cleaner configurations. In your example, it should be the external DHCP server used by the switch network instance where such route is configured.

File: pkg/pillar/nireconciler/linux_config.go (hunk 5 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:827]: `IsMgmt  = ! app-shared`

This if-statement is a negation of that comment (we propagate default route if this if-statement is true).

File: pkg/pillar/nireconciler/linux_config.go (hunk 6 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:845]: This list is also used for static IP routes, which are always propagated (see [here](https://github.com/lf-edge/eve-api/blob/main/proto/config/netinst.proto#L139-L142)).

File: pkg/pillar/nireconciler/genericitems/dnsmasq.go [hunk 7]
```diff
@@ -606,24 +626,57 @@ func (c *DnsmasqConfigurator) CreateDnsmasqConfig(buffer io.Writer, dnsmasq Dnsm
...
+					Gateway:    gatewayIP,
+				})
+			}
+		} else {
+			// With all-ones netmask disabled we use forwarding between apps
+			// on the same NI.
+			if subnet != nil {
+				staticRoutes = append(staticRoutes, types.IPRoute{
+					DstNetwork: subnet,
+					Gateway:    net.IP{0, 0, 0, 0},
```
Comment (naiming-zededa) [inline pkg/pillar/nireconciler/genericitems/dnsmasq.go:652]: why the gateway is 0.0.0.0 here?

File: pkg/pillar/nireconciler/linux_config.go [hunk 8]
```diff
@@ -636,6 +636,40 @@ func (r *LinuxNIReconciler) getIntendedNIL3Cfg(niID uuid.UUID) dg.Graph {
...
+		}
+		family := netlink.FAMILY_V4
+		if route.Gateway.To4() == nil {
+			family = netlink.FAMILY_V6
+		}
+		outputIf := linux.RouteOutIf{UplinkIfName: uplink}
+		if isAppGW {
+			outputIf = linux.RouteOutIf{BridgeIfName: ni.brIfName}
+		}
+		intendedL3Cfg.PutItem(linux.Route{
```
Comment (naiming-zededa) [inline pkg/pillar/nireconciler/linux_config.go:661]: do we check here to prevent the app's static route is 0/0 to be installed into host side?

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/parseconfig.go:441]: But they might not know the external gateway IP a priori whereas they can control the subnet assigned to a network instance. And most network instances have a single uplink.

File: pkg/pillar/nireconciler/linux_config.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/nireconciler/linux_config.go:640]: Ok

File: pkg/pillar/nireconciler/linux_config.go (hunk 4 above)
Comment (eriknordmark) [reply pkg/pillar/nireconciler/linux_config.go:649]: How much documentation are we planning for this feature?

File: pkg/pillar/nireconciler/genericitems/dnsmasq.go (hunk 7 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/genericitems/dnsmasq.go:652]: This is a link-local route for the whole subnet, such as: `10.10.10.0/24 dev eth0 proto dhcp scope link`

File: pkg/pillar/nireconciler/linux_config.go (hunk 8 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:661]: Currently there is no such check. How should we handle the case when NI uplink has default route and at the same time user configures static default route (with possibly different GW IP)? Which one should take precedence? I would think that the static route, since it would give the user the option to override the default route received from an external DHCP server.

File: pkg/pillar/nireconciler/linux_config.go (hunk 4 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:649]: Not much, I will prepare some paragraph for eve docs once we finish this review and clarify all the edge cases.

**Review by eriknordmark — APPROVED** [review]
Kick off tests

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 1 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:441]: Valid point. I will make the necessary changes to allow this.

Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go]: Done

File: pkg/pillar/nireconciler/linux_config.go (hunk 5 above)
Comment (eriknordmark) [reply pkg/pillar/nireconciler/linux_config.go:837]: OK, but I still don't understand why you need to check IsMgmt

Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:837]: For management uplink we always want to propagate default route to apps. For app-shared uplink (IsMgmt=false) only if the uplink has default route.

**Review by eriknordmark — APPROVED** [review]
Run tests

File: pkg/pillar/nireconciler/linux_config.go (hunk 8 above)
Comment (naiming-zededa) [reply pkg/pillar/nireconciler/linux_config.go:671]: Yeh, i would think the user explicit one should take precedence.

Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:684]: Done, PR is updated - user-configured default route will override the default route of the uplink.

File: pkg/pillar/nireconciler/linux_config.go (hunk 5 above)
Comment (eriknordmark) [reply pkg/pillar/nireconciler/linux_config.go:850]: Ah - your terminology might be confusing me.
We have two cases: inject a default route for the app instance (independent of routing on the external network)
propagate a default and/or other routes from the external network to the app instance.

Is it that the same code does the injection (for the management uplinks) and does the propagation?
If so it makes perfect sense, but is hard to understand from reading the code.

Comment (milan-zededa) [reply pkg/pillar/nireconciler/linux_config.go:850]: This code here (`getIntendedDnsmasqCfg` function) only decides if app should receive default route for this network instance or not. If yes, then we propagate `0.0.0.0/0 via bridgeIP` - i.e. in this function we do not care what the actual gateway is and whether it is received from an external network or injected by user.
In function `getIntendedNIL3Cfg` we then prepare config for the NI routing table (for the host network stack). There we decide if we should copy default route from uplink routing table or put there a user-injected default route.

File: pkg/pillar/nireconciler/linux_config.go (hunk 6 above)
Comment (eriknordmark) [reply pkg/pillar/nireconciler/linux_config.go:868]: Ok

File: pkg/pillar/nireconciler/genericitems/dnsmasq.go (hunk 2 above)
Comment (milan-zededa) [reply pkg/pillar/nireconciler/genericitems/dnsmasq.go:673]: Tested with Windows-es - looks good, IP routes are installed as expected.
